bench-tps/src/bench.rs

```diff
@@ -625,42 +586,73 @@ pub fn fund_keys<T: Client>(
                 send_txs.stop();
                 debug!("send {} txs: {}us", to_fund_txs.len(), send_txs.as_us());
 
+                // Sleep a few slots to allow transactions to process
+                sleep(Duration::from_secs(1));
+
                 let mut verify_txs = Measure::start("verify_txs");
-                let mut starting_txs = to_fund_txs.len();
-                let mut verified_txs = 0;
-                let mut failed_verify = 0;
+                let starting_txs = to_fund_txs.len();
+                let verified_txs = Arc::new(AtomicUsize::new(0));
+                let too_many_failures = Arc::new(AtomicBool::new(false));
                 // Only loop multiple times for small (quick) transaction batches
                 for _ in 0..(if starting_txs < 1000 { 3 } else { 1 }) {
-                    let mut timer = Instant::now();
-                    to_fund_txs.retain(|(_, tx)| {
-                        if timer.elapsed() >= Duration::from_secs(5) {
-                            if failed_verify > 0 {
-                                debug!("total txs failed verify: {}", failed_verify);
+                    let failed_verify = Arc::new(AtomicUsize::new(0));
```

Comment:
what do you think about breaking this up into another function. The complexity was bad before, now it's even worse.

Reply:
Totally in favour, I'll clean this up